README.md
# postcss-px-to-viewport

[![NPM version](https://badge.fury.io/js/postcss-px-to-viewport.svg)](http://badge.fury.io/js/postcss-px-to-viewport)

English | [中文](README_CN.md)

A plugin for [PostCSS](https://github.com/postcss/postcss) that generates viewport units (vw, vh, vmin, vmax) from pixel units.

## Demo

If your project involves a fixed width, this script will help to convert pixels into viewport units.

### Input

```css
.class {
  margin: -10px 0.5vh;
  padding: 5vmin 9.5px 1px;
  border: 3px solid black;
  border-bottom-width: 1px;
  font-size: 14px;
  line-height: 20px;
}

.class2 {
  padding-top: 10px; /* px-to-viewport-ignore */
  /* px-to-viewport-ignore-next */
  padding-bottom: 10px;
  /* Any other comment */
  border: 1px solid black;
  margin-bottom: 1px;
  font-size: 20px;
  line-height: 30px;
}

@media (min-width: 750px) {
  .class3 {
    font-size: 16px;
    line-height: 22px;
  }
}
```

### Output

```css
.class {
  margin: -3.125vw 0.5vh;
  padding: 5vmin 2.96875vw 1px;
  border: 0.9375vw solid black;
  border-bottom-width: 1px;
  font-size: 4.375vw;
  line-height: 6.25vw;
}

.class2 {
  padding-top: 10px;
  padding-bottom: 10px;
  /* Any other comment */
  border: 1px solid black;
  margin-bottom: 1px;
  font-size: 6.25vw;
  line-height: 9.375vw;
}

@media (min-width: 750px) {
  .class3 {
    font-size: 16px;
    line-height: 22px;
  }
}
```

## Getting Started

### Installation

Add via npm

```
$ npm install @exclison/postcss-px-to-viewport --save-dev
```

or yarn

```
$ yarn add -D @exclison/postcss-px-to-viewport
```

### Usage

Default Options:

```js
{
  unitToConvert: 'px',
  viewportWidth: 320,
  unitPrecision: 5,
  propList: ['*'],
  viewportUnit: 'vw',
  fontViewportUnit: 'vw',
  selectorBlackList: [],
  minPixelValue: 1,
  mediaQuery: false,
  replace: true,
  exclude: undefined,
  include: undefined,
  landscape: false,
  landscapeUnit: 'vw',
  landscapeWidth: 568
}
```

- `unitToConvert` (String) unit to convert, by default, it is px.
- `viewportWidth` (Number) The width of the viewport.
- `unitPrecision` (Number) The decimal numbers to allow the vw units to grow to.
- `propList` (Array) The properties that can change from px to vw.
  - Values need to be exact matches.
  - Use wildcard _ to enable all properties. Example: ['_']
  - Use * at the start or end of a word. (['*position\*'] will match background-position-y)
  - Use ! to not match a property. Example: ['*', '!letter-spacing']
  - Combine the "not" prefix with the other prefixes. Example: ['*', '!font*']
- `viewportUnit` (String) Expected units.
- `fontViewportUnit` (String) Expected units for font.
- `selectorBlackList` (Array) The selectors to ignore and leave as px.
  - If value is string, it checks to see if selector contains the string.
    - `['body']` will match `.body-class`
  - If value is regexp, it checks to see if the selector matches the regexp.
    - `[/^body$/]` will match `body` but not `.body`
- `minPixelValue` (Number) Set the minimum pixel value to replace.
- `mediaQuery` (Boolean) Allow px to be converted in media queries.
- `replace` (Boolean) replaces rules containing vw instead of adding fallbacks.
- `exclude` (Regexp or Array of Regexp) Ignore some files like 'node_modules'
  - If value is regexp, will ignore the matches files.
  - If value is array, the elements of the array are regexp.
- `include` (Regexp or Array of Regexp) If `include` is set, only matching files will be converted,
  for example, only files under `src/mobile/` (`include: /\/src\/mobile\//`)
  - If the value is regexp, the matching file will be included, otherwise it will be excluded.
  - If value is array, the elements of the array are regexp.
- `landscape` (Boolean) Adds `@media (orientation: landscape)` with values converted via `landscapeWidth`.
- `landscapeUnit` (String) Expected unit for `landscape` option
- `landscapeWidth` (Number) Viewport width for landscape orientation.

> `exclude` and `include` can be set together, and the intersection of the two rules will be taken.

#### Ignoring

You can use special comments for ignore conversion of single lines:

- `/* px-to-viewport-ignore-next */` — on a separate line, prevents conversion on the next line.
- `/* px-to-viewport-ignore */` — after the property on the right, prevents conversion on the same line.

Example:

```css
/* example input: */
.class {
  /* px-to-viewport-ignore-next */
  width: 10px;
  padding: 10px;
  height: 10px; /* px-to-viewport-ignore */
  border: solid 2px #000; /* px-to-viewport-ignore */
}

/* example output: */
.class {
  width: 10px;
  padding: 3.125vw;
  height: 10px;
  border: solid 2px #000;
}
```

There are several more reasons why your pixels may not convert, the following options may affect this:
`propList`, `selectorBlackList`, `minPixelValue`, `mediaQuery`, `exclude`, `include`.

#### Use with PostCss configuration file

add to your `postcss.config.js`

```js
module.exports = {
  plugins: {
    // ...
    'postcss-px-to-viewport': {
      // options
    }
  }
}
```

#### Use with gulp-postcss

add to your `gulpfile.js`:

```js
var gulp = require('gulp')
var postcss = require('gulp-postcss')
var pxtoviewport = require('@exclison/postcss-px-to-viewport')

gulp.task('css', function () {
  var processors = [
    pxtoviewport({
      viewportWidth: 320,
      viewportUnit: 'vmin'
    })
  ]

  return gulp
    .src(['build/css/**/*.css'])
    .pipe(postcss(processors))
    .pipe(gulp.dest('build/css'))
})
```

## Running the tests

In order to run tests, you need to install dev-packages:

```
$ npm install
```

Then run the tests via npm script:

```
$ cd example
$ node index.js
```

## Changelog

The changelog is [here](CHANGELOG.md).


## Authors

- [Exclison](https://github.com/exclison/postcss-px-to-viewport)

See also the list of [contributors](https://github.com/exclison/postcss-px-to-viewport/contributors) who participated in this project.

## License

This project is licensed under the [ISC License](LICENSE).


## Acknowledgments

- Hat tip to https://github.com/evrone/postcss-px-to-viewport for inspiring us for this project.


CHANGELOG.md




## [1.0.0] -2021.09.09

### 完成PX-TO-VIEWPORT主要功能

src/prop-list-matcher.js
const filterPropList = {
  exact: function (list) {
    return list.filter(function (m) {
      return m.match(/^[^\*\!]+$/)
    })
  },
  contain: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\*.+\*$/)
      })
      .map(function (m) {
        return m.substr(1, m.length - 2)
      })
  },
  endWith: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\*[^\*]+$/)
      })
      .map(function (m) {
        return m.substr(1)
      })
  },
  startWith: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^[^\*\!]+\*$/)
      })
      .map(function (m) {
        return m.substr(0, m.length - 1)
      })
  },
  notExact: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\![^\*].*$/)
      })
      .map(function (m) {
        return m.substr(1)
      })
  },
  notContain: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\!\*.+\*$/)
      })
      .map(function (m) {
        return m.substr(2, m.length - 3)
      })
  },
  notEndWith: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\!\*[^\*]+$/)
      })
      .map(function (m) {
        return m.substr(2)
      })
  },
  notStartWith: function (list) {
    return list
      .filter(function (m) {
        return m.match(/^\![^\*]+\*$/)
      })
      .map(function (m) {
        return m.substr(1, m.length - 2)
      })
  }
}

function createPropListMatcher(propList) {
  const hasWild = propList.indexOf('*') > -1
  const matchAll = hasWild && propList.length === 1
  const lists = {
    exact: filterPropList.exact(propList),
    contain: filterPropList.contain(propList),
    startWith: filterPropList.startWith(propList),
    endWith: filterPropList.endWith(propList),
    notExact: filterPropList.notExact(propList),
    notContain: filterPropList.notContain(propList),
    notStartWith: filterPropList.notStartWith(propList),
    notEndWith: filterPropList.notEndWith(propList)
  }
  return function (prop) {
    if (matchAll) return true
    return (
      (hasWild ||
        lists.exact.indexOf(prop) > -1 ||
        lists.contain.some(function (m) {
          return prop.indexOf(m) > -1
        }) ||
        lists.startWith.some(function (m) {
          return prop.indexOf(m) === 0
        }) ||
        lists.endWith.some(function (m) {
          return prop.indexOf(m) === prop.length - m.length
        })) &&
      !(
        lists.notExact.indexOf(prop) > -1 ||
        lists.notContain.some(function (m) {
          return prop.indexOf(m) > -1
        }) ||
        lists.notStartWith.some(function (m) {
          return prop.indexOf(m) === 0
        }) ||
        lists.notEndWith.some(function (m) {
          return prop.indexOf(m) === prop.length - m.length
        })
      )
    )
  }
}

module.exports = {
  filterPropList,
  createPropListMatcher
}


index.js
const postcss = require("postcss");

const { createPropListMatcher } = require("./src/prop-list-matcher");
const { getUnitRegexp } = require("./src/pixel-unit-regexp");

const defaults = {
  unitToConvert: "px", //(String) 需要转换的单位，默认为"px"
  viewportWidth: 320, //(Number) 设计稿的视口宽度
  viewportHeight: 568, //(Number) 设计稿的视口高度
  unitPrecision: 5, //(Number) 单位转换后保留的精度
  viewportUnit: "vw", //(String) 希望使用的视口单位
  fontViewportUnit: "vw", //(String) 字体使用的视口单位
  selectorBlackList: [], //(Array) 需要忽略的CSS选择器
  propList: ["*"], //(Array) 能转化为vw的属性列表
  minPixelValue: 1, //(Number) 设置最小的转换数值，如果为1的话，只有大于1的值会被转换
  mediaQuery: false, //(Boolean) 媒体查询里的单位是否需要转换单位
  replace: true, //(Boolean) 是否直接更换属性值，而不添加备用属性
  landscape: false, //(Boolean) 是否添加根据 landscapeWidth 生成的媒体查询条件 @media (orientation: landscape)
  landscapeUnit: "vw", //(String) 横屏时使用的单位
  landscapeWidth: 568, //(Number) 横屏时使用的视口宽度
};
const ignoreNextComment = "px-to-viewport-ignore-next";
const ignorePrevComment = "px-to-viewport-ignore";
module.exports = (options = {}) => {
  const opts = Object.assign({}, defaults, options);

  checkRegExpOrArray(opts, "exclude");
  checkRegExpOrArray(opts, "include");

  const pxRegex = getUnitRegexp(opts.unitToConvert);
  const satisfyPropList = createPropListMatcher(opts.propList);

  return {
    postcssPlugin: "postcss-px-to-viewport",
    prepare(result) {
      return {
        Once(root, { result }) {
          let landscapeRules = [];
          //遍历规则
          root.walkRules((rule) => {
            const file = rule.source.input.file;

            //包含某些文件
            if (opts.include && file) {
              if (queryTypeString(opts.include) === "[object RegExp]") {
                if (!opts.include.test(file)) return;
              }
              if (queryTypeString(opts.include) === "[object Array]") {
                let flag = false;
                for (let i = 0; i < opts.include.length; i++) {
                  if (opts.include[i].test(file)) {
                    flag = true;
                    break;
                  }
                }
                if (!flag) return;
              }
            }

            //排除某些文件
            if (opts.exclude && file) {
              if (queryTypeString(opts.exclude) === "[object RegExp]") {
                if (opts.exclude.test(file)) return;
              }
              if (queryTypeString(opts.exclude) === "[object Array]") {
                for (let i = 0; i < opts.exclude.length; i++) {
                  if (opts.exclude[i].test(file)) return;
                }
              }
            }

            //过滤选择器
            if (blacklistedSelector(opts.selectorBlackList, rule.selector)) {
              return;
            }

            //如果该rule 是媒体查询下的,并且没有mediaQuery属性,则直接返回,不需要处理
            if (rule.parent.params && !opts.mediaQuery) {
              return;
            }

            //遍历属性
            rule.walkDecls((decl) => {

              //过滤掉不需要操作的属性(即value中不含有px等单位的css样式)
              if (decl.value.indexOf(opts.unitToConvert) < 0) {
                return;
              }
              if (!satisfyPropList(decl.prop)) {
                return;
              }

              const prev = decl.prev();

              //忽略有排除注释的属性,类似忽略eslint类型检查
              if (
                prev &&
                prev.type === "comment" &&
                prev.text === ignoreNextComment
              ) {
                prev.remove();
                return;
              }

              const next = decl.next();

              if (
                next &&
                next.type === "comment" &&
                next.text === ignorePrevComment
              ) {
                if (/\n/.test(next.raws.before)) {
                  result.warn(
                    "Unexpected comment /* " +
                      ignorePrevComment +
                      " */ must be after declaration at same line.",
                    { node: next }
                  );
                } else {
                  next.remove();
                  return;
                }
              }

              let unit; //单位
              let size; //视口宽度
              const params = decl.parent.parent.params;

              if (
                opts.landscape &&
                ((params && params.indexOf("landscape") > 0) || !params)
              ) {
                unit = opts.landscapeUnit;
                size = opts.landscapeWidth;
              } else {
                unit = getUnit(decl.prop, opts);
                size = opts.viewportWidth;
              }

              const value = decl.value.replace(
                pxRegex,
                createPxReplace(opts, unit, size)
              );

              if (declarationExists(decl.parent, decl.prop, value)) {
                return;
              }

              if (opts.replace) {
                decl.value = value;
              } else {
                decl.before(decl.clone({ value: value }));
              }
            });

            if (opts.landscape && !rule.parent.params) {
              const landscapeRule = rule.clone();
              if (landscapeRule.nodes.length > 0) {
                landscapeRules.push(landscapeRule);
              }
            }
          });

          if (landscapeRules.length > 0) {
            const landscapeRoot = postcss.atRule({
              params: "(orientation:landscape)",
              name: "media",
            });
            landscapeRules.forEach((rule) => {
              landscapeRoot.append(rule);
            });
            landscapeRules = [];

            root.append(landscapeRoot);
          }
        },
        Rule(rule) {},
        Declaration(decl, { Rule }) {},
        RuleExit(rule) {},
        RootExit(root) {},
      };
    },
  };
};
module.exports.postcss = true;

//验证参数是否为正则表达式或者正则表达式数组
function checkRegExpOrArray(options, optionName) {
  const option = options[optionName];
  if (!option) return;

  if (queryTypeString(option) === "[object RegExp]") return;
  if (queryTypeString(option) === "[object Array]") {
    const checkRegExp = option.reduce(
      (total, current) =>
        total && queryTypeString(current) === "[object RegExp]",
      true
    );
    if (checkRegExp) return;
  }
  throw new Error(`options ${optionName} should be RegExp or Array of RegExp.`);
}
//获取类型字符串
function queryTypeString(opt) {
  return Object.prototype.toString.call(opt);
}

//验证需要过滤掉的选择器
function blacklistedSelector(blackList, selector) {
  if (typeof selector !== "string") return;

  return blackList.some((regexp) => {
    return typeof regexp === "string"
      ? selector.indexOf(regexp) >= 0
      : selector.match(regexp);
  });
}

//获取单位 正常的单位与文字单位
function getUnit(prop, opts) {
  return prop.indexOf("font") === -1
    ? opts.viewportUnit
    : opts.fontViewportUnit;
}

//px替换为vw
function createPxReplace(opts, viewportUnit, viewportSize) {
  return function (m, $1) {
    if (!$1) return m;
    const pixels = parseFloat($1);
    if (pixels <= opts.minPixelValue) return m;
    const parseVal = toFixed((pixels / viewportSize) * 100, opts.unitPrecision);
    return parseVal === 0 ? "0" : parseVal + viewportUnit;
  };
}
//保留精度
function toFixed(number, precision) {
  const multiplier = Math.pow(10, precision + 1);
  const wholeNumber = Math.floor(number * multiplier);
  return (Math.round(wholeNumber / 10) * 10) / multiplier;
}

//确定是否替换过,替换过的不需要再替换
function declarationExists(decls, prop, value) {
  return decls.some((decl) => decl.prop === prop && decl.value === value);
}
